File: Algorithms/Sorting/quickSort.py
def quickSort(li, start, end):
	if start < end:
		pivot = partition(li, start, end)
		quickSort(li, start, pivot-1)
		quickSort(li, pivot+1, end)
	return li

def partition(li, start, end):
	pivot = li[start]
	left = start+1
	right = end
	finished = False
	while not finished:
		while left <= right and li[left] <= pivot:
			left+=1
		while li[right] >= pivot and right >= left:
			right-=1
		if right < left:
			finished = True
		else:
			temp = li[left]
			li[left] = li[right]
			li[right] = temp
	temp = li[start]
	li[start] = li[right]
	li[right] = temp
	return right


li = [7,2,6,12,98,45,1009,452,192,0,432,879]
print(quickSort(li, 0, len(li)-1))
'''
	prints:
		[0, 2, 6, 7, 12, 45, 98, 192, 432, 452, 879, 1009]
'''

File: Algorithms/Sorting/QuickSort.java
public class QuickSort {
    
    public void quickSort(int[] input, int low, int high) {
        if (low < high) {
            // index found from partition
            int pivot = partition(input, low, high);

            quickSort(input, low, pivot - 1);
            quickSort(input, pivot + 1, high);
        }
    }

    public int partition(int[] input, int low, int high) {
        int pivot = input[high];
        int left = low - 1;

        for (int i = low; i < high; i++) {
            if (input[i] <= pivot) {
                left+=1;

                // swap
                int temp = input[left];
                input[left] = input[i];
                input[i] = temp;
            }
        }

        int temp = input[left+1];
        input[left+1] = input[high];
        input[high] = temp;

        return left + 1;
    }

    public static void main(String[] args) {
        int numbers[] = {22, 4, 16, 891, 506, 2424, 88, 1, 99999, 345, 341, 6792, 8};

        System.out.println("Unsorted");
        for (int i = 0; i < numbers.length; i++) {
            System.out.print(numbers[i] + " ");
        }
        System.out.println("");        

        QuickSort qs = new QuickSort();
        qs.quickSort(numbers, 0, numbers.length-1);

        System.out.println("Sorted");
        for (int i = 0; i < numbers.length; i++) {
            System.out.print(numbers[i] + " ");
        }
    }

}
    
/*
    Running from command line:
        1) javac *.java <--- compile java file to class file
        2) java * <--- Execute compiled Java file
*/